### PR TITLE
Video thumbnail

### PR DESCRIPTION
**Pull request by kevodwyer**

### CODE REVIEW

File: build.xml [hunk 1]
```diff
@@ -207,12 +207,11 @@ Class-Path: ${manifest_cp}
     <copy todir="war/WEB-INF/lib" file="${gwt.sdk}/gwt-servlet-deps.jar" />
     <!-- Add any additional server libs that need to be copied -->
     <copy todir="war/WEB-INF/lib" file="lib/jnr-fuse-0.3-all.jar" />
-    <copy todir="war/WEB-INF/lib" file="lib/orchid-1.0.0.jar" />
   </target>
 
   <target name="javac" depends="libs" description="Compile java source to bytecode">
     <mkdir dir="war/WEB-INF/classes"/>
-    <javac srcdir="src" includes="**" encoding="utf-8"
+    <javac srcdir="src" includes="peergos/client/**" encoding="utf-8"
```
Comment (ianopolous) [inline build.xml]: surely we need peergos/shared as well here?

File: src/peergos/gwt/emu/java/nio/file/StandardOpenOption.java [hunk 2]
```diff
@@ -0,0 +1,7 @@
+package java.nio.file;
+
+public enum StandardOpenOption implements OpenOption {
```
Comment (ianopolous) [inline src/peergos/gwt/emu/java/nio/file/StandardOpenOption.java]: is there a way we can avoid all these superfluous classes? Like have an interface, with a java impl in server, and a native JS one?

File: src/peergos/shared/user/fs/FileTreeNode.java [hunk 3]
```diff
@@ -27,7 +29,7 @@
 public class FileTreeNode {
 
     final static int THUMBNAIL_SIZE = 100;
-
+    private enum FileType { Image, Video, Unknown };
```
Comment (ianopolous) [inline src/peergos/shared/user/fs/FileTreeNode.java]: isn't this covered by the mimetype?

File: src/peergos/shared/user/fs/FileTreeNode.java [hunk 4]
```diff
@@ -302,6 +304,7 @@ private SigningPrivateKeyAndPublicHash getSigner() {
         return ((DirAccess) pointer.fileAccess).getChildrenLocations(pointer.filePointer.baseKey);
     }
 
+    @JsMethod
```
Comment (ianopolous) [inline src/peergos/shared/user/fs/FileTreeNode.java]: what needs this in JS?

File: src/peergos/shared/user/fs/FileTreeNode.java [hunk 5]
```diff
@@ -1003,11 +1006,31 @@ public static FileTreeNode createRoot(TrieNode root) {
         return new byte[0];
     }
 
+    public static byte[] generateVideoThumbnail(byte[] videoBlob) {
```
Comment (ianopolous) [inline src/peergos/shared/user/fs/FileTreeNode.java]: is it possible to do this without creating a random file somewhere? 
also why is .mp4 hard coded?

File: src/peergos/shared/user/fs/FileTreeNode.java [hunk 6]
```diff
@@ -1019,6 +1042,18 @@ public static FileTreeNode createRoot(TrieNode root) {
                             fut.complete(generateThumbnail(bytes));
                         });
                     }
+                }else if(type == FileType.Video) {
+                    if (network.isJavascript()) {
```
Comment (ianopolous) [inline src/peergos/shared/user/fs/FileTreeNode.java]: This is where we could use the interface, say, VideoThumbnailGenerator and put an instance of it in UserContext

File: build.xml (hunk 1 above)
Comment (kevodwyer) [reply build.xml]: it works without it! I don't understand, In any case i will add shared

File: src/peergos/gwt/emu/java/nio/file/StandardOpenOption.java (hunk 2 above)
Comment (kevodwyer) [reply src/peergos/gwt/emu/java/nio/file/StandardOpenOption.java]: Used by shared (FileTreeNode) rather then server. So in short no.

File: src/peergos/shared/user/fs/FileTreeNode.java (hunk 3 above)
Comment (kevodwyer) [reply src/peergos/shared/user/fs/FileTreeNode.java]: not quite as there are many different image/video mimetypes. Using a generic enum rather then using mimeType.startsWith() everywhere seemed like a win.

File: src/peergos/shared/user/fs/FileTreeNode.java (hunk 4 above)
Comment (kevodwyer) [reply src/peergos/shared/user/fs/FileTreeNode.java]: my mistake, fixed (was used in former web-ui branch that is now redundant)

File: src/peergos/shared/user/fs/FileTreeNode.java (hunk 5 above)
Comment (kevodwyer) [reply src/peergos/shared/user/fs/FileTreeNode.java]: javafx Media class can only read from a file. I wish it read from a memory stream. From testing it seemed the file extension was important.

File: src/peergos/gwt/emu/java/nio/file/StandardOpenOption.java (hunk 2 above)
Comment (ianopolous) [reply src/peergos/gwt/emu/java/nio/file/StandardOpenOption.java]: But they're new, so I think we can avoid it by moving the video thumbnail java stuff to a tpye in server.

File: src/peergos/shared/user/fs/FileTreeNode.java (hunk 3 above)
Comment (ianopolous) [reply src/peergos/shared/user/fs/FileTreeNode.java]: But then adding new thumbnails e.g. text, pdf, will all be breaking changes.

File: src/peergos/shared/user/fs/FileTreeNode.java (hunk 5 above)
Comment (ianopolous) [reply src/peergos/shared/user/fs/FileTreeNode.java]: what about .webm or other extensions?

File: src/peergos/shared/user/fs/FileTreeNode.java (hunk 6 above)
Comment (kevodwyer) [reply src/peergos/shared/user/fs/FileTreeNode.java]: The established pattern is to call method on the NativeJSThumbnail class. Return type is string (so can do base64 in gtw translated java), this is in contrast the java method  return byte[]

File: src/peergos/shared/user/fs/FileTreeNode.java (hunk 5 above)
Comment (kevodwyer) [reply src/peergos/shared/user/fs/FileTreeNode.java]: javafx currently only supports limited set of video formats

**Comment by ianopolous** [comment]
Ok this looks good to go, but have you investigated why the tests are timing out?

**Comment by ianopolous** [comment]
Looks like the java video thumbnail test doesn't terminate for some reason?

**Comment by ianopolous** [comment]
Do we need a Platform.exit() now after the test suite?

**Pull request merged**